src/vertex_tools.rs

```diff
@@ -0,0 +1,1272 @@
+use std::f32;
+
+use nalgebra;
+use nalgebra::{Point2, Point3, Rotation3, Vector3};
+
+use crate::convert::cast_usize;
+use crate::geometry::{self, Face, Geometry, Plane, UnorientedEdge};
+
+/// Compute barycentric coordinates of point P in triangle A, B, C. Returns None
+/// for degenerate triangles.
+pub fn compute_barycentric_coords(
+    a: Point2<f32>,
+    b: Point2<f32>,
+    c: Point2<f32>,
+    p: Point2<f32>,
+) -> Option<Point3<f32>> {
+    let ab = b - a;
+    let ac = c - a;
+    let pa = a - p;
+    let xs = Vector3::new(ac.x, ab.x, pa.x);
+    let ys = Vector3::new(ac.y, ab.y, pa.y);
+    let ortho = xs.cross(&ys);
+    if f32::abs(ortho.z) < 1.0 {
+        None
+    } else {
+        Some(Point3::new(
+            1.0 - (ortho.x + ortho.y) / ortho.z,
+            ortho.y / ortho.z,
+            ortho.x / ortho.z,
+        ))
+    }
+}
+
+/// Checks if all three points of a triangle lay on the same line.
+pub fn are_triangle_vertices_colinear(
+    triangle_vertices: (&Point3<f32>, &Point3<f32>, &Point3<f32>),
+) -> bool {
+    let v0_normalized = triangle_vertices.0.coords.normalize();
+    let v1_normalized = triangle_vertices.1.coords.normalize();
+    let v2_normalized = triangle_vertices.2.coords.normalize();
+    v0_normalized == v1_normalized && v0_normalized == v2_normalized
+}
+
+/// Checks if a point lies in a triangle.
+///
+/// #Panics
+/// Panics if triangle is colinear
+///
+/// https://math.stackexchange.com/questions/4322/check-whether-a-point-is-within-a-3d-triangle
+// TODO: Orient the triangle and point to XY plane and then test
+fn is_point_in_triangle(
+    point: &Point3<f32>,
+    triangle_vertices: (&Point3<f32>, &Point3<f32>, &Point3<f32>),
+) -> bool {
+    // If the triangle is degenerated into a point, check if the point is equal
+    // to the test point.
+    if triangle_vertices
+        .0
+        .coords
+        .relative_eq(&triangle_vertices.1.coords, 0.001, 0.001)
+        && triangle_vertices
+            .0
+            .coords
+            .relative_eq(&triangle_vertices.2.coords, 0.001, 0.001)
+    {
+        return point == triangle_vertices.0;
+    }
+
+    // If the triangle is degenerated into a line, check if the point lies on
+    // the line.
+    if are_triangle_vertices_colinear(triangle_vertices) {
+        return is_point_on_line_clamped(point, triangle_vertices.0, triangle_vertices.1);
+    }
+
+    let plane = Plane::from_three_points(
+        triangle_vertices.0,
+        triangle_vertices.1,
+        triangle_vertices.2,
+    );
+
+    // If the point is not on the triangle plane, it's also not on the triangle.
+    if !is_point_on_plane(point, &plane) {
+        return false;
+    }
+
+    // In case the triangle isn't horizontal
+    if let Some(rotation_to_horizontal) =
```

Comment:
Let's merge this `if` branch with early return with the code that follows it. They are almost exact copy-paste.

The overall structure can be something like:

```rust
// pseudocode

let (rotated_triangle, rotated_point) = if check_horizontal(triangle) {
    (triangle, point)
} else {
    if let Some(rotation_to_horizontal) = rotation_between(...) {
        (apply_rotation(triangle, rotation_to_horizontal), apply_rotation(point, rotation_to_horizontal))
    } else {
        // This probably won't happen, but we shouldn't crash in case it does :)
        // https://docs.rs/nalgebra/0.19.0/src/nalgebra/geometry/rotation_specialization.rs.html#663
        (triangle, point)
    }
}

// ... the rest of the code that's the same for both branches, will operate on `rotated_triangle` and `rotated_point`
```

Reply:
I need to apologize. You were right this covers all the cases, however I made the last else statement differently. It should never occur, as you correctly expect, but if it does, it means we're not able to calculate any valid result. In fact we should return None, but that's not helpful for the caller. Therefore i decided to early return false instead. The main reason for the early return is, however, that the barycentric coords coming from this branch won't be valid and therefore the calculations won't yield any valid results either.

Reply:
Ok, let's keep it as is.